Modules/posixmodule.c

```diff
@@ -1341,49 +1343,50 @@ path_converter(PyObject *o, void *p)
 
     length = PyBytes_GET_SIZE(bytes);
     narrow = PyBytes_AS_STRING(bytes);
-    if ((size_t)length != strlen(narrow)) {
+    if (!path->null_embeddable && strlen(narrow) != (size_t)length) {
         FORMAT_EXCEPTION(PyExc_ValueError, "embedded null character in %s");
         goto error_exit;
     }
 
-#ifdef MS_WINDOWS
-    wo = PyUnicode_DecodeFSDefaultAndSize(
-        narrow,
-        length
-    );
-    if (!wo) {
-        goto error_exit;
-    }
+    if (path->make_wide) {
+        wo = PyUnicode_DecodeFSDefaultAndSize(narrow, length);
+        if (!wo) {
+            goto error_exit;
+        }
 
-    wide = PyUnicode_AsWideCharString(wo, &length);
-    Py_DECREF(wo);
-    if (!wide) {
-        goto error_exit;
-    }
-    if (length > 32767) {
-        FORMAT_EXCEPTION(PyExc_ValueError, "%s too long for Windows");
-        goto error_exit;
-    }
-    if (wcslen(wide) != length) {
-        FORMAT_EXCEPTION(PyExc_ValueError, "embedded null character in %s");
-        goto error_exit;
-    }
-    path->wide = wide;
-    path->narrow = TRUE;
-    Py_DECREF(bytes);
-    wide = NULL;
-#else
-    path->wide = NULL;
-    path->narrow = narrow;
-    if (bytes == o) {
-        /* Still a reference owned by path->object, don't have to
-           worry about path->narrow is used after free. */
+        wide = PyUnicode_AsWideCharString(wo, &length);
+        Py_DECREF(wo);
+        if (!wide) {
+            goto error_exit;
+        }
+#ifdef MS_WINDOWS
+        if (length > 32767) {
+            FORMAT_EXCEPTION(PyExc_ValueError, "%s too long for Windows");
+            goto error_exit;
+        }
+#endif
+        if (!path->null_embeddable && wcslen(wide) != (size_t)length) {
+            FORMAT_EXCEPTION(PyExc_ValueError,
+                             "embedded null character in %s");
+            goto error_exit;
+        }
+        path->wide = wide;
+        path->narrow = narrow; // TODO: replace by NULL
```

Comment:
We should replace this by NULL, but we'll have to update the rest of the code accordingly.

```suggestion
        path->narrow = NULL;
```

Reply:
@zooba, could you help me determine which conditions need to be updated?

Reply:
Some code needs to know whether the original object was `bytes` or `str`, to construct an appropriate return value. It could check the type of `object`, but for efficiency there should be a boolean value such as `is_bytes_object`. Then if `wide` is used and `is_bytes_object` is non-zero, code in some cases would know to return a `bytes` result. Or if `narrow` is used and `is_bytes_object` is 0, code in some cases would know to return an `str` result. 

With `path_t` being generalized, the existing checks of `narrow` on Windows will have to be updated to instead check `is_bytes_object`. On POSIX, checks of the type of `object` could also be updated for efficiency.

The following functions check the type of `object` to determine the result type:

* `_posix_listdir`
* `os_readlink_impl`
* `DirEntry_from_posix_info`

The following functions check `narrow` to determine the result type:

* `_listdir_windows_no_opendir`
* `os__getfullpathname_impl`
* `os__getfinalpathname_impl`
* `os__getvolumepathname_impl`
* `os_readlink_impl`
* `DirEntry_from_find_data`

Also, `path_and_dir_fd_invalid` doesn't need to special case Windows.

Reply:
Thanks, I'm going to be a bit careful with adding extra fields to `path_t`. We can always optimise this later.